SQL_SELECT_HW

skalantaryan opened the pull request.

krabts commented on SQL_SELECT_Sona\_Kalantaryan_HW.sql:
```diff
@@ -0,0 +1,119 @@
+--1-1
+SELECT f.title, f.release_year, f.rating
+FROM public.film f
```
This is not mistake, but leave only movie title.

krabts commented on SQL_SELECT_Sona\_Kalantaryan_HW.sql:
```diff
@@ -0,0 +1,119 @@
+ORDER BY f.title ASC;
+
+--1-2
+SELECT a.address || a.address2 AS full_address,
+       SUM(p.amount) AS revenue
+FROM public.store s
+JOIN public.address a ON s.address_id = a.address_id
+JOIN public.inventory i ON s.store_id = i.store_id
+JOIN public.rental r ON i.inventory_id = r.inventory_id
+JOIN public.payment p ON r.rental_id = p.rental_id
+WHERE r.rental_date > '2017-03-31'
+GROUP BY a.address, a.address2
+ORDER BY revenue DESC;
```
1. Unfortunately, I got the following result:
    ![image](https://github.com/user-attachments/assets/4cae2c16-23a9-4628-89b4-8333dfa4d810)

    Could you please show my screen with your result?

2. Please use separate filters for Year and Month here:
```SQL
       WHERE r.rental_date > '2017-03-31'
```

3. Please use store_id first of all in GROUP BY. Different stores could have the same address. You need revenue for stores.

4. You need some separator between addresses.
```SQL
       SELECT a.address || a.address2 AS full_address,
```

Please pay attention that separator (comma, etc.) should be presented only if address2 in not null.

krabts commented on SQL_SELECT_Sona\_Kalantaryan_HW.sql:
```diff
@@ -0,0 +1,119 @@
+LIMIT 5;
+
+--1-4
+SELECT f.release_year,
+       COUNT(CASE WHEN c.name = 'Drama' THEN f.film_id END) AS number_of_drama_movies,
+       COUNT(CASE WHEN c.name = 'Travel' THEN f.film_id END) AS number_of_travel_movies,
+       COUNT(CASE WHEN c.name = 'Documentary' THEN f.film_id END) AS number_of_documentary_movies
+FROM public.film f
+JOIN public.film_category fc ON f.film_id = fc.film_id
+JOIN public.category c ON fc.category_id = c.category_id
+GROUP BY f.release_year
+ORDER BY f.release_year DESC;
```
Please consider such cases when movie has category like `DRAMA`, `drama`, `Drama` by accident.

krabts commented on SQL_SELECT_Sona\_Kalantaryan_HW.sql:
```diff
@@ -0,0 +1,119 @@
+           ROW_NUMBER() OVER (PARTITION BY staff_id ORDER BY last_payment_date DESC) AS store_rank
+    FROM employee_revenue
+)
+SELECT e.first_name, e.last_name, te.total_revenue
+FROM top_employees te
+JOIN public.staff e ON te.staff_id = e.staff_id
+JOIN public.store s ON te.store_id = s.store_id
+WHERE te.store_rank = 1  -- Take the last store where the employee worked
+ORDER BY te.total_revenue DESC
+LIMIT 3;
```
Please concatenate name as full name (is not mistake) and don't forget to add store_id to final result.

krabts reviewed (changes requested):
Hi Sona,

thank you for the good job. Please find my comments below. HW is to be updated. don't merge this.


1. Please define JOIN types explicitly in all queries.
2. Please add some comments in code about most important filters, etc. Comments existence is a part of your score.
3. The rest remarks are in code.

If you have questions or objections, please let me know.

Regards,
Kiryl

krabts commented:
Hi Sona, please add the link to Learn. Also, could you please leave only one sql file in your branch? What exactly should I check? ;)

![image](https://github.com/user-attachments/assets/36b90f9b-c0d8-4bae-a9f1-1d359142c7f3)

krabts commented on SQL_SELECT_Sona\_Kalantaryan_HW.sql:
```diff
@@ -0,0 +1,151 @@
+INNER JOIN public.category c ON fc.category_id = c.category_id
+WHERE c.name = 'Animation' --filter based on the film category, choose only Animation
+AND f.release_year BETWEEN 2017 AND 2019 --filter based on the release year
+AND f.rating IN ('G', 'PG', 'PG-13', 'R') --filter based on the ratings that are higher than 1 - means exclude NC17
+ORDER BY f.title ASC;
```
Maybe I forgot to pay your attention by accidence. This is wrong filter. You should use rental rate, not rating
![image](https://github.com/user-attachments/assets/4eb5e29b-6d4b-4c09-8377-5a024ec271d3)

`filter based on the ratings that are higher than 1 - means exclude NC17` honestly, I don't understand how did you make such conclusion

skalantaryan replied:
As it said rate, I thought it meant the column rate. The other one was called rental_rate. I was not sure about that.

krabts commented on SQL_SELECT_Sona\_Kalantaryan_HW.sql:
```diff
@@ -0,0 +1,151 @@
+ORDER BY f.title ASC;
+
+--1-2
+SELECT 
+	CASE WHEN a.address is NULL OR a.address2 is NULL THEN a.address || a.address2
+		ELSE a.address || ', ' || a.address2 
+	END AS full_address,
+	--when any of the addresses is null then write them without space or , 
+	--if a.address is null it will write only a.address2 and vice vers
+	--and if none of them is null - ELSE - then it will write a.address, a.address2
+    SUM(p.amount) AS revenue
+FROM public.store s
+LEFT JOIN public.address a ON s.address_id = a.address_id --LEFT as the store can not have address, but we will still need it
+INNER JOIN public.inventory i ON s.store_id = i.store_id --INNER as if the store is not in inventory then it does not have payments
+INNER JOIN public.rental r ON i.inventory_id = r.inventory_id --INNER as if the inventory is not in rental it does not have payments
+LEFT JOIN public.payment p ON r.rental_id = p.rental_id --LEFT as the rental can not have payments
+WHERE EXTRACT(YEAR FROM r.rental_date) = 2017 --the movie was rented in 2017
+	AND EXTRACT(MONTH FROM r.rental_date) > 3 --the movie was rented after March
+GROUP BY s.store_id, a.address, a.address2
+ORDER BY revenue DESC;
```
As I said last time, your query doesn't provide any result. Did you run this?
![image](https://github.com/user-attachments/assets/f3e07111-eb9b-47c1-beba-e569005de34c)

1. The issue is that you should use payment date for filtering, not rental date.
2. If you change this, you can get NULL address. Probably you should adjust your statement for address. Thy this as an option
```SQL
CONCAT(ad.address, COALESCE(', ' || ad.address2, '')) AS address
```

krabts commented on SQL_SELECT_Sona\_Kalantaryan_HW.sql:
```diff
@@ -0,0 +1,151 @@
+ORDER BY first_name, last_name;
+
+--3-2
+WITH movie_gaps AS (
+    SELECT a.actor_id, a.first_name, a.last_name, f1.release_year AS release_year,
+           MIN(f2.release_year) AS next_release_year, --this gets the next movie release year
+		   CASE 
+		   	WHEN MIN(f2.release_year) IS NULL
+			   THEN EXTRACT(YEAR FROM CURRENT_DATE) - f1.release_year
+           	ELSE 
+			   MIN(f2.release_year) - f1.release_year 
+			END AS gap_between_films
+			--if there was no next movie then the gap should be calculated till current year
+    FROM public.actor a
+    INNER JOIN public.film_actor fa1 ON a.actor_id = fa1.actor_id
+    INNER JOIN public.film f1 ON fa1.film_id = f1.film_id
+	--INNER JOINs to get actors who have acted in a film
+    LEFT JOIN public.film_actor fa2 ON a.actor_id = fa2.actor_id
+    LEFT JOIN public.film f2 ON fa2.film_id = f2.film_id AND f2.release_year > f1.release_year --we get next movie from release_year 
+    --LEFT JOINs to get even those actors that don't have a next movie
+	GROUP BY a.actor_id, a.first_name, a.last_name, f1.release_year
+	ORDER BY actor_id, first_name, last_name, release_year
+)
+SELECT first_name, last_name, MAX(gap_between_films) AS longest_gap
+FROM movie_gaps
+WHERE gap_between_films IS NOT NULL
+GROUP BY first_name, last_name
+HAVING MAX(gap_between_films) = (SELECT MAX(gap_between_films) FROM movie_gaps) 
+--choose the ones that have the largest gap between movies
+ORDER BY first_name, last_name;
```
Why this task provides the same result as task 3.1? Last time it was different.

![image](https://github.com/user-attachments/assets/9127a9c5-8cc5-433e-9171-bc7aa87200b6)

skalantaryan replied:
I realized I had used a window function so I changed it.
And while writing the new version I thought I could calculate the gap between current year and the date of the last release year if the actor didn't have a next movie, as it is also considered a gap.
Is it wrong?

krabts replied:
Task 3.2 should provide different result. Your query provides the same as 3.1.

krabts reviewed (changes requested):
Hi Sona,

thank you for updated. Unfortunately, I found weird things. Please check my comments. If you don't agree with something, please let me know.

Be attentive, please :)

Regards,
Kiryl

krabts commented on SQL_SELECT_Sona\_Kalantaryan_HW.sql:
```diff
@@ -0,0 +1,151 @@
+
+
+--2-1
+WITH employee_revenue AS (
+    SELECT s.staff_id,
+           s.store_id,
+           SUM(p.amount) AS total_revenue,
+           MAX(p.payment_date) AS last_payment_date
+    FROM public.staff s
+    LEFT JOIN public.payment p ON s.staff_id = p.staff_id
+	--LEFT JOIN as the staff can not have payment
+    INNER JOIN public.rental r ON p.rental_id = r.rental_id
+	--INNER JOIN as if the staff doesn not have rental we don't need it 
+    WHERE EXTRACT(YEAR FROM p.payment_date) = 2017
+    GROUP BY s.staff_id, s.store_id
+),--we get the each staff's total amount of revenue in each store
+	top_employees AS (
+    SELECT staff_id, store_id, total_revenue,
+           ROW_NUMBER() OVER (PARTITION BY staff_id ORDER BY last_payment_date DESC) AS store_rank 
+		   --This will assign ranks(row numbers) by grouping rows by last_payment_date, 
+		   --so that we can choose the last store the staff worked in
+    FROM employee_revenue
+)--we give rank to each store based on order of working places of each staff
+SELECT s.store_id, e.first_name || ' ' || e.last_name, te.total_revenue
+FROM top_employees te
+INNER JOIN public.staff e ON te.staff_id = e.staff_id
+INNER JOIN public.store s ON te.store_id = s.store_id
+--INNER JOINs as we need each employee's revenue from each store
+WHERE te.store_rank = 1  --choose the last place of work of each employee
+ORDER BY te.total_revenue DESC;
```
Why did you get this time 5 employees? 

![image](https://github.com/user-attachments/assets/e062a43d-2adc-4b80-b683-b9c8b992638f)

Last time your solution had only 3 as required.

![image](https://github.com/user-attachments/assets/c1a89a90-0f5d-4b8a-bc1c-8aa32d577c68)

skalantaryan replied:
I don't know what I did wrong but I added LIMIT 3.

krabts replied:
Honestly, I don't understand your explanation. Task is to provide 3 employees, isn't it? You provided 5.

krabts commented on SQL_SELECT_Sona\_Kalantaryan_HW.sql:
```diff
@@ -0,0 +1,151 @@
+WHERE years_since_last_movie = (SELECT MAX(years_since_last_movie) FROM actor_last_movie)
+ORDER BY first_name, last_name;
+
+--3-2
+WITH movie_gaps AS (
+    SELECT a.actor_id, a.first_name, a.last_name, f1.release_year AS release_year,
+           MIN(f2.release_year) AS next_release_year, --this gets the next movie release year
+		   CASE 
+		   	WHEN MIN(f2.release_year) IS NULL
+			   THEN EXTRACT(YEAR FROM CURRENT_DATE) - f1.release_year
+           	ELSE 
+			   MIN(f2.release_year) - f1.release_year 
+			END AS gap_between_films
+			--if there was no next movie then the gap should be calculated till current year
+    FROM public.actor a
+    INNER JOIN public.film_actor fa1 ON a.actor_id = fa1.actor_id
+    INNER JOIN public.film f1 ON fa1.film_id = f1.film_id
```
This is what you provided first time

![image](https://github.com/user-attachments/assets/bb44e450-190f-49df-8736-5702bfcd2fab)

It should be improved.

This time your 3.2 task has the same result as 3.1.

![image](https://github.com/user-attachments/assets/32684a56-038a-4033-bbc5-376dab2ca777)

If I am wrong, please explain me.

skalantaryan replied:
In the first version it says calculate the gap as from the last film till current year. And in the second version it says to calculate gap between consecutive films. I don't know if I am right or no, but I thought that I should also calculate gaps between current year if there is no next movie. If there is no next movie it is also a gap - that's what I thought. But if it is not required in the second version, I will remove that line.

krabts replied:
I can repeat. Your first solution contained these two lines:
![image](https://github.com/user-attachments/assets/1c0b5a29-cec5-4e93-842c-ebcee5f2ccdf)

Is this your first solution 3.2?

krabts reviewed (changes requested):
Hi Sona,

1. Provide the link in Learn
2. As I said, you need only one HW file in your branch
    ![image](https://github.com/user-attachments/assets/d79ccfd4-79fa-43c1-8beb-fce952696a2f)
3. Please check my other comments.

Regards,
Kiryl

skalantaryan commented:
> Hi Sona,
> 
> 1. Provide the link in Learn
> 2. As I said, you need only one HW file in your branch
>    ![image](https://private-user-images.githubusercontent.com/157591026/428372155-d79ccfd4-79fa-43c1-8beb-fce952696a2f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDMzMTIyODgsIm5iZiI6MTc0MzMxMTk4OCwicGF0aCI6Ii8xNTc1OTEwMjYvNDI4MzcyMTU1LWQ3OWNjZmQ0LTc5ZmEtNDNjMS04YmViLWZjZTk1MjY5NmEyZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMzMFQwNTE5NDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04ZjlmZmUzMjc2MzljNzdkZjhlMjJiNDIwOTgwYzdhY2Q3NWViYmIwZTE2ZGU5Mjc0NGEzNGRjODBmYzIwM2U4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.nBrMvL4yD3AefSnxE6pwthqQKeuQruY6dUA0UJiyDz8)
> 3. Please check my other comments.
> 
> Regards, Kiryl

I just added the file with the wrong name the first time so when I noticed I added the new one with the required name,  and I hadn't seen that the first one was not removed.
I will delete it now.

krabts reviewed (approved):
Hi Sona,

thank you for the changes made. This PR has been approved. You can merge.
Good luck with the next HW ;)

Regards,
Kiryl
